File: Revision/src/main/resources/static/js/sudokuSolver.js

function isSafe(grid,row,col,n,position){
    for (let i = 0; i < n; i++) {
        if(grid[row][i]===position)
            return false;
        if(grid[i][col]===position)
            return false;
    }
    //Checking in the sudoku square
    const boxLength = math.sqrt(n);
    const rowSize = row - row%boxLength;
    const colSize = col - col%boxLength;

    for(let i=0;i<rowSize;i++){
        for (let j = 0; j < colSize; j++) {
            if(grid[boxLength+rowSize][boxLength+colSize]===position)
                return false;
        }
    }
    return true;

}

function sudokuSolver(grid){
    const n = grid.length;
    let row=-1,col=-1,flag=false;
    for(let i=0;i<n;i++){
        for(let j=0;j<n;j++){
            if(grid[i][j]===0){
                flag = true;
                break;
            }
            if(!flag)
                break;


        }

    }
    if(flag)
        return true;

    for(let i=1;i<=n;i++){
        grid[row][col]=i;
        if(isSafe(grid,row,col,n,i)) {
            if (sudokuSolver(grid))
                return true;
            else
                grid[row][col] = 0;
        }
    }

}

File: Algorithms/src/main/java/com/cleo/algorithms/backTracking/SudokuSolver.java
package com.cleo.algorithms.backTracking;

import java.util.ArrayList;

public class SudokuSolver {

    public boolean isSafe(ArrayList<ArrayList<Integer>> board, int row, int col, int n, int N){
        for (int i = 0; i < N; i++) {
            if(board.get(row).get(i)==n||board.get(i).get(col)==n)
                return false;
        }
        int grid1= (int)(Math.sqrt(N));
        int rs = row - row % grid1;
        int cs = col - col % grid1;
        for (int i = 0; i < grid1; i++) {
            for (int j = 0; j < grid1; j++) {
                if (board.get(i + rs).get(j + cs) == n)
                    return false;
            }

        }
        return true;


    }

    public boolean solve(ArrayList<ArrayList<Integer>> board, int N){
        int row=0,col=0;
        boolean isEmpty=false;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if(board.get(i).get(j)==0){
                    row=i;col=j;
                    isEmpty=true;
                    break;
                }
            }
            if(isEmpty)
                break;
        }
        if(!isEmpty)
            return true;

        for (int i = 1; i <= N; i++) {
            if(isSafe(board,row,col,i,N)){
                board.get(row).set(col,i);
                if(solve(board,N))
                    return true;
                else
                    board.get(row).set(col,0);
            }
        }
            return false;
    }

    public ArrayList<ArrayList<Integer>> solveSudoku(ArrayList<ArrayList<Integer>> board) {
        // Write your code here.
       solve(board,board.size());
       return board;
    }

    public static void main(String[] args) {

    }
}
